### PR TITLE
fix backwards compatibility breakages from 1.21 api changes

### LINKED ISSUE #5851: IncompatibleClassChangeError
### Type of bug

Error in console

### `/ess dump all` output

https://essentialsx.net/dump.html?id=930bdb12556e4f079ddf5b6a693966df

### Error log (if applicable)

https://paste.gg/p/anonymous/387f18ae3d2b454d92833042ec0e07a6

### Bug description

What happens when you click on any inventory item

### Steps to reproduce

just click item

### Expected behaviour

This all happens in versions 1.20.4 and lower.

### Actual behaviour

X

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by CyberFlameGO**
### Information

This PR fixes #5851. 

### Details

**Proposed fix:**    
Get the top inventory via reflection, and get the bottom inventory via getting clicked inventory rather than via view.

**Environments tested:**    

#### Will test later on and populate these fields - just drafting the PR for review

OS: 

Java version: 

- [ ] Most recent Paper version (1.XX.Y, git-Paper-BUILD)
- [ ] CraftBukkit/Spigot/Paper 1.12.2
- [ ] CraftBukkit 1.8.8

**Demonstration:**    
N/A - if this PR will work then there is nothing to screenshot other than lack of exceptions in logs

### CODE REVIEW

File: Essentials/src/main/java/com/earth2me/essentials/EssentialsPlayerListener.java [hunk 1]
```diff
@@ -894,17 +893,28 @@ public void run() {
         return used;
     }
 
+    public static Inventory getTopInventory(InventoryEvent event) {
+        try {
+            Object view = event.getView();
+            Method getTopInventory = view.getClass().getMethod("getTopInventory");
```
Comment (mbax) [inline Essentials/src/main/java/com/earth2me/essentials/EssentialsPlayerListener.java:899]: Expensive reflection should be done once. Just call invoke in this method.

Comment (CyberFlameGO) [reply Essentials/src/main/java/com/earth2me/essentials/EssentialsPlayerListener.java:899]: Like this? Or would it be better to solely declare the variable outside the method and change the declaration inside the method to an assignment
```java
    private static Method getTopInventoryMethod;

    static {
        try {
            getTopInventoryMethod = InventoryView.class.getMethod("getTopInventory");
            getTopInventoryMethod.setAccessible(true);
        } catch (NoSuchMethodException e) {
            throw new RuntimeException(e);
        }
    }

    public static Inventory getTopInventory(InventoryEvent event) {
        try {
            Object view = event.getView();
            return (Inventory) getTopInventoryMethod.invoke(view);
        } catch (InvocationTargetException | IllegalAccessException e) {
            throw new RuntimeException(e);
        }
    }
```

Comment (mbax) [reply Essentials/src/main/java/com/earth2me/essentials/EssentialsPlayerListener.java:899]: I'd shorten further by just calling `invoke(event.getView())`

Comment (zlataovce) [reply Essentials/src/main/java/com/earth2me/essentials/EssentialsPlayerListener.java:899]: does the method need to have its accessibility explicitly set? it should be public and is fetched with `getMethod`, which fetches only public members - it should already be accessible

**Review by JRoy — CHANGES_REQUESTED** [review]
blocking till I get a chance to properly review

**Comment by JRoy** [comment]
I do not think reflection is the way to go here, especially considering that some of this code could run multiple times per tick. (And if we did we did go the reflection route I'd want to use cached MethodHandles).

I think the best route here is using a provider, I will open a PR with this shortly. Thanks for looking into this anyway.

**Pull request closed without merging**